### PR TITLE
feat: batect-wrapper manager

### PR DESCRIPTION
**Pull request by charleskorn**
## Changes:

This PR adds support for updating the Batect wrapper scripts (`batect` and `batect.cmd`).

## Context:

This is the second of three PRs to implement what is described in #7891. (See #7898 for the first PR.)

The Batect wrapper scripts are two shell scripts (one for macOS / Linux and another for Windows) that are committed alongside the codebase, similar to the Gradle wrapper (`gradlew` / `gradlew.bat`). 

A simple string replacement of the old version number with the new version number is not enough - the wrapper script content can change from version to version. The wrapper scripts are published as assets on the corresponding GitHub release, so the manager must download these and commit them as part of the PR.

https://github.com/batect/batect-sample-ruby/commit/0bd6efb453546c51f80f6c2303c791958dec924d is an example of the simplest possible update, and https://github.com/batect/batect-sample-ruby/commit/fe6860e7c05a6692a3a92d7e9471f36bc0a1ed54 is an example of a larger change.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [x] Unit tests + ran on a real repository

Can be tested against [`charleskorn/kaml`](https://github.com/charleskorn/kaml).

### CODE REVIEW

File: lib/manager/batect-wrapper/extract.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  const dependency: PackageDependency = {
+    depName: 'batect/batect',
+    commitMessageTopic: 'Batect',
+    currentValue: match.groups.version,
+    datasource: githubReleaseDatasource,
+    versioning: semverVersioning,
+  };
+
+  logger.debug(dependency, 'Found Batect wrapper version');
```
Comment (viceice) [inline lib/manager/batect-wrapper/extract.ts:26]:
```suggestion
  logger.trace(dependency, 'Found Batect wrapper version');
```

File: lib/manager/batect-wrapper/extract.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { id as githubReleaseDatasource } from '../../datasource/github-releases';
+import { logger } from '../../logger';
+import { id as semverVersioning } from '../../versioning/semver';
+import { PackageDependency, PackageFile } from '../common';
+
+const VERSION_REGEX = /^\s+VERSION="(?<version>.*)"$/gm;
+
+export function extractPackageFile(fileContent: string): PackageFile | null {
+  logger.debug('batect.extractPackageFile()');
+  const matches = [...fileContent.matchAll(VERSION_REGEX)];
```
Comment (viceice) [inline lib/manager/batect-wrapper/extract.ts:10]: Why `matchAll` can't we simply search for first match and ignore others?

File: lib/manager/batect-wrapper/artifacts.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        name: packageFileName,
+        contents: await getFileContent(release, 'batect'),
+      },
+    },
+    {
+      file: {
+        name: `${packageFileName}.cmd`,
+        contents: await getFileContent(release, 'batect.cmd'),
+      },
+    },
```
Comment (viceice) [inline lib/manager/batect-wrapper/artifacts.ts:66]: Didn't need the `batect.jar` file be updated too?

**Review by viceice — CHANGES_REQUESTED** [review]
I think we don't need to change the datasource. We can simply direct download known files:

`https://github.com/batect/batect/releases/download/0.64.0/batect.cmd`

We just need the new version. So please revert datasource changes and use direct download url

**Comment by rarkins** [comment]
In that case I agree, now that there is a simple way to access both the files. @charleskorn can you refactor as proposed by @viceice ?

File: lib/manager/batect-wrapper/artifacts.ts (hunk 3 above)
Comment (charleskorn) [reply lib/manager/batect-wrapper/artifacts.ts]: No - the `.jar` file isn't committed to the repository, the wrapper script downloads it to the user's computer on first run.

**Comment by charleskorn** [comment]
Thanks for all the feedback, I've incorporated the suggested changes.

File: lib/manager/batect-wrapper/extract.ts (hunk 1 above)
Comment (viceice) [reply lib/manager/batect-wrapper/extract.ts:24]: I think you missed this?

File: lib/manager/batect-wrapper/artifacts.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  fileName: string
+): Promise<string> {
+  const url = `https://github.com/batect/batect/releases/download/${version}/${fileName}`;
+
+  try {
+    const response = await http.get(url);
+
+    return response.body;
+  } catch (err) {
+    const errorDescription: string = err.toString();
```
Comment (viceice) [inline lib/manager/batect-wrapper/artifacts.ts:18]: Move try catch to updateArtifacts and return artifacts error instead throw. Checkout other manager like gradle wrapper.

File: lib/manager/batect-wrapper/extract.ts (hunk 1 above)
Comment (charleskorn) [reply lib/manager/batect-wrapper/extract.ts:24]: Good catch, thanks

**Comment by charleskorn** [comment]
PR build seems to have failed randomly after the merge and I can't re-run it - but otherwise I think this is good for another review.

**Review by rarkins — APPROVED** [review]

**Pull request merged**